src/lib/hooks/swap/useWrapCallback.tsx

```diff
@@ -0,0 +1,132 @@
+import { ContractTransaction } from '@ethersproject/contracts'
+import { Trans } from '@lingui/macro'
+import { useWETHContract } from 'hooks/useContract'
+import { useAtomValue } from 'jotai/utils'
+import { Field, swapAtom } from 'lib/state/swap'
+import tryParseCurrencyAmount from 'lib/utils/tryParseCurrencyAmount'
+import { useCallback, useMemo, useState } from 'react'
+
+import { WRAPPED_NATIVE_CURRENCY } from '../../../constants/tokens'
+import useActiveWeb3React from '../useActiveWeb3React'
+import { useCurrencyBalances } from '../useCurrencyBalance'
+import useNativeCurrency from '../useNativeCurrency'
+
+export enum WrapInputError {
+  NO_ERROR, // must be equal to 0 so all other errors are truthy
+  ENTER_NATIVE_AMOUNT,
+  ENTER_WRAPPED_AMOUNT,
+  INSUFFICIENT_NATIVE_BALANCE,
+  INSUFFICIENT_WRAPPED_BALANCE,
+}
+
+export function WrapErrorText({ wrapInputError }: { wrapInputError: WrapInputError }) {
+  const native = useNativeCurrency()
+  const wrapped = native?.wrapped
+
+  switch (wrapInputError) {
+    case WrapInputError.ENTER_NATIVE_AMOUNT:
+      return <Trans>Enter {native?.symbol} amount</Trans>
+    case WrapInputError.ENTER_WRAPPED_AMOUNT:
+      return <Trans>Enter {wrapped?.symbol} amount</Trans>
+    case WrapInputError.INSUFFICIENT_NATIVE_BALANCE:
+      return <Trans>Insufficient {native?.symbol} balance</Trans>
+    case WrapInputError.INSUFFICIENT_WRAPPED_BALANCE:
+      return <Trans>Insufficient {wrapped?.symbol} balance</Trans>
+    case WrapInputError.NO_ERROR:
+    default:
+      return null
+  }
+}
+
+export enum WrapType {
+  NOT_APPLICABLE,
+  WRAP,
+  UNWRAP,
+}
+interface UseWrapCallbackReturns {
+  callback: () => Promise<ContractTransaction>
+  error: WrapInputError
+  loading: boolean
+  type: WrapType
+}
+
+export default function useWrapCallback(): UseWrapCallbackReturns {
+  const { account, chainId } = useActiveWeb3React()
+  const [loading, setLoading] = useState(false)
+  const wethContract = useWETHContract()
+  const {
+    amount,
+    independentField,
+    [Field.INPUT]: inputCurrency,
+    [Field.OUTPUT]: outputCurrency,
+  } = useAtomValue(swapAtom)
+
+  const wrapType = useMemo(() => {
+    if (!inputCurrency || !outputCurrency || !chainId) {
+      return WrapType.NOT_APPLICABLE
+    }
+    const weth = WRAPPED_NATIVE_CURRENCY[chainId]
```

Comment:
should we change this from `weth` as we support more chains without ETH as native currency? (MATIC) for example

Reply:
I think so. This is the pattern currently used in app, so I was a little hesitant to change.